### DISCUSSION TITLE
Polymorphism and Polymorphic Models/Tables

### DISCUSSION THREAD

**Discussion by marcotas** [adonisjs/core#3801] [Polls]
Since adonis is inspired by Laravel framework I miss a very common use case on all applications I've worked which is polymorphic relationships/tables/models the same way [Laravel does](https://laravel.com/docs/9.x/eloquent-relationships#polymorphic-relationships). So I'd like to suggest a polymorphic relationships and tables with Adonisjs. Here are some models/tables very commonly used with polymorphic relations:

- tags or categories. You can basically have many different models that are taggable: in a CRM you can add tags to contacts, attachments, organizations, emails etc.
- Comments: you can add comments to many relations: threads, channel, post, vídeo, image etc.
- attachments or files: you can also have many models with many different type of attachments but you would like to have one implementation to manage them all. [Media library](https://spatie.be/docs/laravel-medialibrary/v10/introduction) is a really good example of a use case for multiple attachments (files) in different models.
- Addresses: in many applications, you might want to have addresses of users, contacts, customers and organizations.

These are all common and already used use cases that I've worked with using Laravel and it was just amazing. I really would love to use it in Adonisjs. ❤️ 

By the way, congratulations for this awesome framework 🚀 ❤️

**Comment by Salmondx** [comment]
I agree, that is a great feature to have and in some cases it is almost impossible to avoid it.

Imagine that you have a storefront that points to objects of different types that have different properties. With polymorphism you would just have a single `StoreItem` or whatever entity that points to different product types. Store item would have properties shared by every object it points to, such as thumbnail, search keywords, description, name, etc.

Without polymorphism how would you deal with such scenario? Duplicate same properties for each distinct type you have in a system? How would you paginate results then? How to add new field that is shared across different types? Create N different migrations for each product type? How to make a unified API that points to a store front item, instead of building routes for each product type? And lot's of other problems.

 Polymorphism have it's cons for sure such as having to do N extra queries for each type that is being fetched. But in this case developers just need to understand it and limit the amount of distinct types they use with polymorphic tables. Polymorphism goes against the relational database normalization theory but lots of modern things do as well. Having `json` columns were considered an insane anti-pattern 10 years ago but nowadays everyone does it because it is really convenient in some cases and most RDMS even support search functionality for such columns.

Personally I don't see why Lucid shouldn't have such functionality while both Laravel and Rails have it. I'm using such pattern on a relatively high-load product with Laravel with a table that have ~3 million rows in it and 3-5 different types of objects it points to. Haven't seen any significant problems or latency issues with it on RDS instances with SSD.

@thetutlage @RomainLanz sorry for pinging you, but maybe you could share your opinion about this problem and how would you solve issues I listed above without using polymorphism? Is there a better way that I'm not aware of? I've seen Harminder's responses from 2017-2018 but he didn't mention any specific concerns and may be he changed his opinion since then?

**Comment by MohammedElattar** [comment]
That would be very nice but i think the core team members are not interested about this approach, so i found a workaround to handle it 

https://github.com/adonisjs/lucid/issues/86#issuecomment-2784748661

**Reply by bretoreta** [reply]
Then I think calling AdonisJS laravel like is just to sell the idea, they are not interested in making it "like" laravel at all. 

I have been wanting this for so long as well, but I'm stuck.
